pkg/sql/catalog/tabledesc/logical_replication_helpers.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tabledesc
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/errors"
+)
+
+const cannotLDRMsg = "cannot create logical replication stream"
+
+// CheckLogicalReplicationCompatibility verifies that the destination table
+// descriptor is a valid target for logical replication from the source table.
+func CheckLogicalReplicationCompatibility(src, dst *descpb.TableDescriptor) error {
+	if err := checkSrcDstColsMatch(src, dst); err != nil {
+		return err
+	}
+
+	if err := checkColumnFamilies(dst); err != nil {
+		return err
+	}
+
+	if err := checkCompositeTypesInPrimaryKey(dst); err != nil {
+		return err
+	}
+
+	if err := checkVirtualColumns(dst); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// We disallow expression evaluation (e.g. virtual columns that are stored in an
+// index) because the LDR KV write path does not understand how to evaluate
+// expressions. The writer expects to receive the full set of columns, even the
+// computed ones, along with a list of columns that we've already determined
+// should be updated.
+func checkVirtualColumns(dst *descpb.TableDescriptor) error {
+	// Disallow partial indexes.
+	for _, idx := range dst.Indexes {
+		if idx.IsPartial() {
+			return errors.WithMessage(pgerror.Newf(
+				pgcode.InvalidTableDefinition,
+				"table %s has a partial index %s", dst.Name, idx.Name,
+			), cannotLDRMsg)
+		}
+	}
+
```

Comment:
i think we also need to check if any secondary index contains a storing clause that contains a virtual column

Reply:
check the comment right below this. :) 
```
NB: it is impossible for a virtual column to be stored in an index.
```